File: sorting/quickSort.js
// Time complexity: O(nlog2(n)) average, O(n^2) worst case
// Space complexity: O(log2(n))
const quickSort = (array) => {
  return quickSortHelper(array, 0, array.length - 1);
};

const quickSortHelper = (array, low, high) => {
  const index = partition(array, low, high);
  if (low < index - 1) quickSortHelper(array, low, index - 1);

  if (index < high) quickSortHelper(array, index, high);
  return array;
};

const partition = (array, low, high) => {
  const pivot = array[Math.floor((low + high) / 2)];
  while (low <= high) {
    while (array[low] < pivot) low++;
    while (array[high] > pivot) high--;
    if (low <= high) {
      const temp = array[low];
      array[low] = array[high];
      array[high] = temp;
      low++;
      high--;
    }
  }
  return low;
};

File: sorting/quickSort.py
from typing import List

def quickSort(nums: List[int]) -> List[int]:
  n = len(nums)
  sort(nums, 0, n - 1)
  return nums

def sort(nums: List[int], low: int, high: int) -> List[int]:
  if low >= high: return
  index = partition(nums, low, high)
  sort(nums, low, index - 1)
  sort(nums, index + 1, high)

def partition(nums: List[int], low: int, high: int) -> int:
  mid = (low + high) // 2
  pivot = nums[mid]
  nums[high], nums[mid] = nums[mid], nums[high]
  i = low 
  for j in range(low, high):
    if nums[j] < pivot:
      nums[i], nums[j] = nums[j], nums[i]
      i += 1
  nums[i], nums[high] = nums[high], nums[i]
  return i